storage/transfermanager/downloader.go

```diff
@@ -69,6 +73,82 @@ func (d *Downloader) DownloadObject(ctx context.Context, input *DownloadObjectIn
 	return nil
 }
 
+// DownloadDirectory queues the download of a set of objects to a local path.
+// This will initiate the download but is non-blocking; call Downloader.Results
+// or use the callback to process the result. DownloadDirectory is thread-safe
+// and can be called simultaneously from different goroutines.
+// DownloadDirectory will resolve any filters on the input and create the needed
+// directory structure locally as the operations progress.
+// Note: DownloadDirectory overwrites existing files in the directory.
+func (d *Downloader) DownloadDirectory(ctx context.Context, input *DownloadDirectoryInput) error {
+	if d.config.asynchronous && input.Callback == nil {
+		return errors.New("transfermanager: input.Callback must not be nil when the WithCallbacks option is set")
+	}
+	if !d.config.asynchronous && input.Callback != nil {
+		return errors.New("transfermanager: input.Callback must be nil unless the WithCallbacks option is set")
+	}
+
+	select {
+	case <-d.doneReceivingInputs:
+		return errors.New("transfermanager: WaitAndClose called before DownloadDirectory")
+	default:
+	}
+
+	objectsToQueue := []DownloadObjectInput{}
+	query := &storage.Query{
+		Prefix:      input.Prefix,
+		StartOffset: input.StartOffset,
+		EndOffset:   input.EndOffset,
+		MatchGlob:   input.MatchGlob,
+	}
+	if err := query.SetAttrSelection([]string{"Name"}); err != nil {
+		return fmt.Errorf("transfermanager: DownloadDirectory query.SetAttrSelection: %w", err)
+	}
+
+	it := d.client.Bucket(input.Bucket).Objects(ctx, query)
+
+	// TODO: Clean up any created directory structure on failure.
+
+	attrs, err := it.Next()
+	for err != iterator.Done {
```

Comment:
I feel like this is a slightly confusing way to write this loop; is there a reason you do it this way vs. how we write it in the list objects sample?

Reply:
Honestly this seems more intuitive to me, but I can change it so it's consistent with the samples since it's usually how we do it.

Reply:
Yeah I would prefer that, it just seems odd to have the first call of Next happen before the loop.